File: Logic Building Problems/1_solution_even_or_odd.cpp
#include <iostream>
using namespace std;

bool isEvenNum(int number) {
    /*
        By checking wheather the remainder is zero or not.
    */
    if (number % 2 == 0) {
        return true;
    } else {
        return false;
    }
}

bool isEven(int number) {
    /*
        There is not any operation in this method.
        It just checks the bit of the number 
        for the even number the bit is always 0 and
        for odd the bit is 1 
    */
    if ((number & 1) == 0) {
        return true;
    } else {
        return false;
    }
}



int main() {
    int input_number;
    cin >> input_number;
    if (isEven(input_number)) {
        cout << "true" << endl;
    } else {
        cout << "false" << endl;
    }
    return 0;
}

File: Logic_Building_Problems/1_solution_even_or_odd.py
def check_even_or_odd_with_remainder(number):
    if number % 2 == 0:
        return True
    else:
        return False
    
def check_even_or_odd_with_bitwise(number):
    """
        This method is more efficient because it does not do any operation,
        just check the bit of a number. For the even number the bit is always
        0 and for the odd number the bit is 1. That's it.
    """
    if number & 1 == 0:
        return True
    else:
        return False
        

if __name__ == '__main__':
    number = int(input())
    if check_even_or_odd_with_bitwise(number=number):
        print("true")
    else:
        print("false")